File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/map/ConstrainMode.java
package org.maplibre.nativeffi.map;

/** Constraint mode used by map viewport options. */
public final class ConstrainMode {
  public static final ConstrainMode NONE = new ConstrainMode(0);
  public static final ConstrainMode HEIGHT_ONLY = new ConstrainMode(1);
  public static final ConstrainMode WIDTH_AND_HEIGHT = new ConstrainMode(2);
  public static final ConstrainMode SCREEN = new ConstrainMode(3);

  private final int rawValue;
  private final String name;

  public ConstrainMode(int rawValue) {
    this.rawValue = rawValue;
    this.name =
        switch (rawValue) {
          case 0 -> "NONE";
          case 1 -> "HEIGHT_ONLY";
          case 2 -> "WIDTH_AND_HEIGHT";
          case 3 -> "SCREEN";
          default -> "UNKNOWN(" + Integer.toUnsignedLong(rawValue) + ")";
        };
  }

  public int rawValue() {
    return rawValue;
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof ConstrainMode that && rawValue == that.rawValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(rawValue);
  }

  @Override
  public String toString() {
    return name;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/map/ConstrainMode.kt
package org.maplibre.nativeffi.map

import kotlin.jvm.JvmInline

/**
 * Constraint mode used by map viewport options.
 *
 * This is an open domain: MapLibre Native may report a value that has no named constant here, so a
 * `when` over this type needs an `else` branch. Unknown values are preserved as their raw
 * [nativeValue] rather than collapsed to a known constant.
 */
@JvmInline
public value class ConstrainMode(public val nativeValue: Int) {
  public companion object {
    public val NONE: ConstrainMode = ConstrainMode(0)
    public val HEIGHT_ONLY: ConstrainMode = ConstrainMode(1)
    public val WIDTH_AND_HEIGHT: ConstrainMode = ConstrainMode(2)
    public val SCREEN: ConstrainMode = ConstrainMode(3)

    internal fun fromNative(nativeValue: UInt): ConstrainMode = fromNative(nativeValue.toInt())

    internal fun fromNative(nativeValue: Int): ConstrainMode = ConstrainMode(nativeValue)
  }

  internal val isKnown: Boolean
    get() = this == NONE || this == HEIGHT_ONLY || this == WIDTH_AND_HEIGHT || this == SCREEN
}
